nemo_aligner/utils/trt_llm.py

```diff
@@ -55,32 +70,44 @@ def __init__(
         self.unload_engine_train = unload_engine_train
         self.trt_model_type = trt_model_type
         self.reshard_model = reshard_model
-        self.pad_id = tokenizer.eos_id
-
         self.trt_llm_exporter = TensorRTLLM(trt_model_dir, load_model=False)
         self._trtllm_model_compiled = False
 
         rng_generator = torch.Generator(device="cpu")
         seed = secrets.randbits(32) if seed is None else seed
         rng_generator.manual_seed(seed)
-
         self.rng_generator = rng_generator
 
+        self.pad_id = tokenizer.pad_id
+        end_id = tokenizer.eos_id
         end_strings = list(end_strings)
-        assert all(
-            "," not in end_string for end_string in end_strings
-        ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
 
-        end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+        # TRT-LLM uses different logic to compute the response length depending on if we specify stop_list or end_id
+        # we want to use the same logic(which is to include the stop token in response length) so we put end_id into the stop_list
+        # manually
+        if len(end_strings) == 0:
+            ids, offsets = [end_id], [1]
+        else:
+            assert all(
+                "," not in end_string for end_string in end_strings
+            ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
+
+            end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+            # use an arbitary ref string to obtain end_string ids
+            stop_list = tensorrt_llm_run.to_word_list_format(
+                end_strings, build_tokenizer(self.tokenizer), ref_str="green tea icecream"
+            )
+            # add the end_id, we only need to manipulate one slice
+            ids, offsets = stop_list[0].tolist()
+            ids = append_and_repad_list(ids, end_id, pad_id=0)
+            offsets = append_and_repad_list(offsets, max(offsets) + 1, pad_id=-1)
 
-        # use an arbitary ref string to obtain end_string ids
-        stop_list = tensorrt_llm_run.to_word_list_format(
-            end_strings, build_tokenizer(self.tokenizer), ref_str="green tea icecream"
+        stop_list = torch.as_tensor([ids, offsets], dtype=torch.int32, device=torch.cuda.current_device()).repeat(
+            self.generation_batch_size, 1, 1
         )
-        stop_list = torch.from_numpy(stop_list).cuda().contiguous()
 
         self.sampling_config = tensorrt_llm.runtime.SamplingConfig(
-            end_id=tokenizer.eos_id,
+            end_id=self.pad_id,
```

Comment:
What would happen if we set `end_id=-1`? (if it works could we get rid of the constraint `tokenizer.pad_id != tokenizer.eos_id`?)

Reply:
i tried this and when i set end_id=-1, all the padding tokens become -1 so i don't think it works